File: skills/academic-figure/templates/rf-0027-baseline-table/python/plot.py
#!/usr/bin/env python3
"""Render a schema-driven descriptive baseline table."""

from __future__ import annotations

import argparse
import csv
import math
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

import matplotlib

matplotlib.use("Agg")
matplotlib.rcParams["svg.fonttype"] = "none"

import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle
import numpy as np


MAX_GROUP_COLUMNS = 6
MAX_DISPLAY_ROWS = 40


class ContractError(ValueError):
    pass


@dataclass(frozen=True)
class VariableSpec:
    variable: str
    label: str
    kind: str
    levels: tuple[str, ...]
    summary: str
    decimals: int


@dataclass(frozen=True)
class TableRow:
    kind: str
    label: str
    cells: tuple[str, ...]


def ordered_unique(values: Iterable[str]) -> tuple[str, ...]:
    return tuple(dict.fromkeys(values))


def read_schema(path: Path) -> tuple[VariableSpec, ...]:
    if not path.is_file():
        raise ContractError(f"schema file does not exist: {path}")
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        required = ["variable", "label", "type", "levels", "summary", "decimals"]
        if reader.fieldnames is None or any(name not in reader.fieldnames for name in required):
            raise ContractError("schema must contain: " + ", ".join(required))
        if len(reader.fieldnames) != len(set(reader.fieldnames)):
            raise ContractError("schema contains duplicate column names")
        specs: list[VariableSpec] = []
        for row_number, row in enumerate(reader, start=2):
            variable = (row.get("variable") or "").strip()
            label = (row.get("label") or "").strip()
            kind = (row.get("type") or "").strip()
            levels_text = (row.get("levels") or "").strip()
            summary = (row.get("summary") or "").strip()
            decimals_text = (row.get("decimals") or "").strip()
            if not variable or not label:
                raise ContractError(f"schema row {row_number}: variable and label are required")
            if variable in {"id", "group"}:
                raise ContractError(f"schema row {row_number}: reserved variable {variable!r}")
            if kind not in {"continuous", "categorical"}:
                raise ContractError(f"schema row {row_number}: type must be continuous or categorical")
            try:
                decimals = int(decimals_text)
            except ValueError as error:
                raise ContractError(f"schema row {row_number}: decimals must be an integer") from error
            if decimals < 0 or decimals > 6:
                raise ContractError(f"schema row {row_number}: decimals must be between 0 and 6")
            levels = tuple(part.strip() for part in levels_text.split("|") if part.strip())
            if kind == "continuous":
                if levels:
                    raise ContractError(f"schema row {row_number}: continuous levels must be empty")
                if summary not in {"mean_sd", "median_iqr"}:
                    raise ContractError(f"schema row {row_number}: continuous summary must be mean_sd or median_iqr")
            else:
                if not levels or len(levels) != len(set(levels)):
                    raise ContractError(f"schema row {row_number}: categorical levels must be unique and pipe-delimited")
                if summary != "n_percent_nonmissing":
                    raise ContractError(f"schema row {row_number}: categorical summary must be n_percent_nonmissing")
            specs.append(VariableSpec(variable, label, kind, levels, summary, decimals))
    if not specs:
        raise ContractError("schema contains no variables")
    if len({spec.variable for spec in specs}) != len(specs):
        raise ContractError("schema variable names must be unique")
    return tuple(specs)


def read_data(path: Path, specs: tuple[VariableSpec, ...]) -> list[dict[str, object]]:
    if not path.is_file():
        raise ContractError(f"input file does not exist: {path}")
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ContractError("input must contain a header row")
        if len(reader.fieldnames) != len(set(reader.fieldnames)):
            raise ContractError("input contains duplicate column names")
        required = ["id", "group", *[spec.variable for spec in specs]]
        missing = [name for name in required if name not in reader.fieldnames]
        if missing:
            raise ContractError("input is missing schema field(s): " + ", ".join(missing))
        records: list[dict[str, object]] = []
        errors: list[str] = []
        for row_number, row in enumerate(reader, start=2):
            subject_id = (row.get("id") or "").strip()
            group = (row.get("group") or "").strip()
            if not subject_id or not group:
                errors.append(f"row {row_number}: id and group must be non-empty")
                continue
            record: dict[str, object] = {"id": subject_id, "group": group}
            for spec in specs:
                text = (row.get(spec.variable) or "").strip()
                if text == "":
                    record[spec.variable] = None
                elif spec.kind == "continuous":
                    try:
                        value = float(text)
                    except ValueError:
                        errors.append(f"row {row_number}: {spec.variable} is not numeric")
                        value = math.nan
                    if not math.isfinite(value):
                        errors.append(f"row {row_number}: {spec.variable} must be finite")
                    record[spec.variable] = value
                else:
                    if text not in spec.levels:
                        errors.append(f"row {row_number}: {spec.variable}={text!r} is not declared in schema levels")
                    record[spec.variable] = text
            records.append(record)
    if errors:
        raise ContractError("input violates the typed schema:\n  - " + "\n  - ".join(errors[:10]))
    if not records:
        raise ContractError("input contains no subjects")
    ids = [str(record["id"]) for record in records]
    if len(ids) != len(set(ids)):
        raise ContractError("id must be unique: one input row per subject")
    return records


def group_columns(records: list[dict[str, object]], include_overall: bool):
    order = ordered_unique(str(record["group"]) for record in records)
    columns = [(group, [record for record in records if record["group"] == group]) for group in order]
    if include_overall:
        if "Overall" in order:
            raise ContractError("group label 'Overall' conflicts with --include-overall true")
        columns.append(("Overall", records))
    if len(columns) > MAX_GROUP_COLUMNS:
        raise ContractError(
            f"table would have {len(columns)} group columns; maximum is {MAX_GROUP_COLUMNS}. "
            "Split the table using a scientifically defined grouping."
        )
    return columns


def number(value: float, decimals: int) -> str:
    return f"{value:.{decimals}f}"


def build_rows(specs: tuple[VariableSpec, ...], columns) -> tuple[TableRow, ...]:
    rows: list[TableRow] = []
    for spec in specs:
        availability = []
        for _, records in columns:
            available = sum(record[spec.variable] is not None for record in records)
            availability.append(f"available {available}/{len(records)}; missing {len(records) - available}")
        rows.append(TableRow("section", spec.label, tuple(availability)))
        if spec.kind == "continuous":
            label = "Mean (SD)" if spec.summary == "mean_sd" else "Median [Q1, Q3]"
            cells = []
            for _, records in columns:
                values = np.asarray(
                    [float(record[spec.variable]) for record in records if record[spec.variable] is not None],
                    dtype=float,
                )
                if not values.size:
                    cells.append("NA")
                elif spec.summary == "mean_sd":
                    sd_text = number(float(np.std(values, ddof=1)), spec.decimals) if values.size > 1 else "NA"
                    cells.append(f"{number(float(np.mean(values)), spec.decimals)} ({sd_text})")
                else:
                    q1, median, q3 = np.quantile(values, (0.25, 0.5, 0.75), method="linear")
                    cells.append(
                        f"{number(float(median), spec.decimals)} "
                        f"[{number(float(q1), spec.decimals)}, {number(float(q3), spec.decimals)}]"
                    )
            rows.append(TableRow("data", label, tuple(cells)))
        else:
            for level in spec.levels:
                cells = []
                for _, records in columns:
                    available = [record[spec.variable] for record in records if record[spec.variable] is not None]
                    count = sum(value == level for value in available)
                    percent = 100.0 * count / len(available) if available else math.nan
                    cells.append(f"{count}/{len(available)} ({percent:.1f}%)" if available else "0/0 (NA)")
                rows.append(TableRow("data", level, tuple(cells)))
    if len(rows) > MAX_DISPLAY_ROWS:
        raise ContractError(
            f"schema expands to {len(rows)} display rows; maximum is {MAX_DISPLAY_ROWS}. "
            "Split variables into themed tables rather than shrinking text."
        )
    return tuple(rows)


def draw_table(title: str, columns, rows: tuple[TableRow, ...]) -> plt.Figure:
    maximum_label = max(len(row.label) for row in rows)
    width = min(20.0, max(8.5, 4.6 + 2.1 * len(columns) + 0.045 * maximum_label))
    height = max(5.0, 2.4 + 0.42 * (len(rows) + 1))
    figure = plt.figure(figsize=(width, height), facecolor="white")
    axis = figure.add_axes([0, 0, 1, 1])
    axis.set_xlim(0, 1)
    axis.set_ylim(0, 1)
    axis.axis("off")
    colors = {
        "ink": "#17212B", "muted": "#617080", "header": "#17354D",
        "section": "#EAF3F2", "stripe": "#F7F9FA", "line": "#D8E0E5", "white": "#FFFFFF"
    }
    left, right, top, bottom = 0.035, 0.975, 0.865, 0.105
    label_fraction = min(0.44, max(0.30, 0.28 + 0.004 * maximum_label))
    label_right = left + (right - left) * label_fraction
    data_width = (right - label_right) / len(columns)
    edges = [left, label_right] + [label_right + data_width * index for index in range(1, len(columns) + 1)]
    units = [1.28] + [1.05 if row.kind == "section" else 1.0 for row in rows]
    unit_height = (top - bottom) / sum(units)

    axis.text(left, 0.962, title, ha="left", va="top", fontsize=18, fontweight="semibold", color=colors["ink"])
    axis.text(
        left, 0.920,
        f"{sum(len(records) for label, records in columns if label != 'Overall')} subject rows · "
        f"{len(columns)} displayed group columns · descriptive summaries only",
        ha="left", va="top", fontsize=9.5, color=colors["muted"]
    )
    y = top
    header_height = units[0] * unit_height
    axis.add_patch(Rectangle((left, y - header_height), right - left, header_height, color=colors["header"], ec="none"))
    axis.text(left + 0.012, y - header_height / 2, "Variable / summary", ha="left", va="center", fontsize=10, fontweight="bold", color="white")
    for index, (label, records) in enumerate(columns):
        center = (edges[index + 1] + edges[index + 2]) / 2
        axis.text(center, y - header_height / 2, f"{label}\nN={len(records)}", ha="center", va="center", fontsize=9.3, fontweight="bold", color="white")
    y -= header_height
    data_index = 0
    for row, weight in zip(rows, units[1:]):
        row_height = weight * unit_height
        fill = colors["section"] if row.kind == "section" else colors["stripe"] if data_index % 2 else colors["white"]
        axis.add_patch(Rectangle((left, y - row_height), right - left, row_height, color=fill, ec="none"))
        axis.plot([left, right], [y - row_height, y - row_height], color=colors["line"], lw=0.65)
        axis.text(
            left + (0.012 if row.kind == "section" else 0.022), y - row_height / 2, row.label,
            ha="left", va="center", fontsize=9.2, fontweight="bold" if row.kind == "section" else "normal", color=colors["ink"]
        )
        for index, cell in enumerate(row.cells):
            center = (edges[index + 1] + edges[index + 2]) / 2
            axis.text(center, y - row_height / 2, cell, ha="center", va="center", fontsize=7.9 if row.kind == "section" else 8.8, color=colors["muted"] if row.kind == "section" else colors["ink"])
        if row.kind == "data":
            data_index += 1
        y -= row_height
    axis.add_patch(Rectangle((left, bottom), right - left, top - bottom, fill=False, ec=colors["header"], lw=1.0))
    axis.text(
        left, 0.060,
        "Categorical percentages use each variable's non-missing denominator. Continuous summaries exclude missing values. "
        "SD is sample SD; quartiles use linear interpolation. No P values or SMD are computed.",
        ha="left", va="top", fontsize=8.1, color=colors["muted"]
    )
    return figure


def report(records, specs, columns) -> None:
    missing_cells = sum(record[spec.variable] is None for record in records for spec in specs)
    print(
        f"Loaded {len(records)} unique subject rows: {len(columns)} displayed group column(s), "
        f"{len(specs)} typed variable(s), {missing_cells} missing value cell(s); 0 subject rows excluded."
    )
    for spec in specs:
        print(f"variable={spec.variable} | type={spec.kind} | summary={spec.summary}")
        for label, group_records in columns:
            available = sum(record[spec.variable] is not None for record in group_records)
            print(f"  group={label} | N={len(group_records)} | available={available} | missing={len(group_records)-available}")


def parse_bool(value: str) -> bool:
    if value.lower() not in {"true", "false"}:
        raise argparse.ArgumentTypeError("expected true or false")
    return value.lower() == "true"


def main() -> int:
    parser = argparse.ArgumentParser(description="Render a schema-driven descriptive baseline table.")
    parser.add_argument("--input", required=True, type=Path)
    parser.add_argument("--schema", required=True, type=Path)
    parser.add_argument("--output-prefix", required=True, type=Path)
    parser.add_argument("--title", default="Baseline characteristics")
    parser.add_argument("--include-overall", type=parse_bool, default=False)
    args = parser.parse_args()
    if args.output_prefix.suffix.lower() in {".png", ".svg"}:
        print("ERROR: --output-prefix must not include .png or .svg", file=sys.stderr)
        return 2
    try:
        specs = read_schema(args.schema)
        records = read_data(args.input, specs)
        columns = group_columns(records, args.include_overall)
        rows = build_rows(specs, columns)
    except ContractError as error:
        print(f"ERROR: {error}", file=sys.stderr)
        return 2
    figure = draw_table(args.title, columns, rows)
    args.output_prefix.parent.mkdir(parents=True, exist_ok=True)
    png_path = Path(f"{args.output_prefix}.png")
    svg_path = Path(f"{args.output_prefix}.svg")
    figure.savefig(png_path, dpi=320, facecolor="white", bbox_inches="tight")
    figure.savefig(svg_path, facecolor="white", bbox_inches="tight")
    plt.close(figure)
    report(records, specs, columns)
    print(f"Wrote {png_path}")
    print(f"Wrote {svg_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/academic-figure/templates/rf-0027-baseline-table/r/plot.R
#!/usr/bin/env Rscript

MAX_GROUP_COLUMNS <- 6L
MAX_DISPLAY_ROWS <- 40L

fail <- function(text) stop(text, call. = FALSE)
ordered_unique <- function(values) values[!duplicated(values)]

parse_args <- function(args) {
  config <- list(
    input = NULL,
    schema = NULL,
    output_prefix = NULL,
    title = "Baseline characteristics",
    include_overall = "false"
  )
  index <- 1L
  while (index <= length(args)) {
    token <- args[[index]]
    if (!startsWith(token, "--")) fail(paste0("unexpected argument: ", token))
    body <- substring(token, 3L)
    if (grepl("=", body, fixed = TRUE)) {
      pieces <- strsplit(body, "=", fixed = TRUE)[[1L]]
      key <- pieces[[1L]]
      value <- paste(pieces[-1L], collapse = "=")
    } else {
      key <- body
      if (index == length(args)) fail(paste0("missing value for --", key))
      index <- index + 1L
      value <- args[[index]]
    }
    key <- gsub("-", "_", key, fixed = TRUE)
    if (!key %in% names(config)) fail(paste0("unknown option: --", body))
    config[[key]] <- value
    index <- index + 1L
  }
  for (required in c("input", "schema", "output_prefix")) {
    if (is.null(config[[required]]) || !nzchar(config[[required]])) fail(paste0("--", gsub("_", "-", required), " is required"))
  }
  if (grepl("\\.(png|svg)$", config$output_prefix, ignore.case = TRUE)) {
    fail("--output-prefix must not include .png or .svg")
  }
  if (!tolower(config$include_overall) %in% c("true", "false")) fail("--include-overall must be true or false")
  config$include_overall <- tolower(config$include_overall) == "true"
  config
}

read_schema <- function(path) {
  if (!file.exists(path)) fail(paste0("schema file does not exist: ", path))
  schema <- read.csv(
    path,
    header = TRUE,
    check.names = FALSE,
    stringsAsFactors = FALSE,
    colClasses = "character",
    na.strings = character(),
    fileEncoding = "UTF-8-BOM"
  )
  required <- c("variable", "label", "type", "levels", "summary", "decimals")
  if (!all(required %in% names(schema))) fail(paste0("schema must contain: ", paste(required, collapse = ", ")))
  if (anyDuplicated(names(schema))) fail("schema contains duplicate column names")
  if (!nrow(schema)) fail("schema contains no variables")
  for (field in required) schema[[field]] <- trimws(schema[[field]])
  if (any(!nzchar(schema$variable)) || any(!nzchar(schema$label))) fail("schema variable and label are required")
  if (any(schema$variable %in% c("id", "group"))) fail("schema uses reserved variable id or group")
  if (anyDuplicated(schema$variable)) fail("schema variable names must be unique")
  if (any(!schema$type %in% c("continuous", "categorical"))) fail("schema type must be continuous or categorical")
  decimals <- suppressWarnings(as.integer(schema$decimals))
  if (any(is.na(decimals)) || any(as.character(decimals) != schema$decimals) || any(decimals < 0L | decimals > 6L)) {
    fail("schema decimals must be integers between 0 and 6")
  }
  schema$decimals <- decimals
  schema$level_list <- lapply(schema$levels, function(text) {
    if (!nzchar(text)) return(character())
    values <- trimws(strsplit(text, "|", fixed = TRUE)[[1L]])
    values[nzchar(values)]
  })
  for (index in seq_len(nrow(schema))) {
    if (schema$type[[index]] == "continuous") {
      if (length(schema$level_list[[index]])) fail("continuous schema levels must be empty")
      if (!schema$summary[[index]] %in% c("mean_sd", "median_iqr")) fail("continuous summary must be mean_sd or median_iqr")
    } else {
      levels <- schema$level_list[[index]]
      if (!length(levels) || anyDuplicated(levels)) fail("categorical levels must be unique and pipe-delimited")
      if (schema$summary[[index]] != "n_percent_nonmissing") fail("categorical summary must be n_percent_nonmissing")
    }
  }
  schema
}

read_data <- function(path, schema) {
  if (!file.exists(path)) fail(paste0("input file does not exist: ", path))
  data <- read.csv(
    path,
    header = TRUE,
    check.names = FALSE,
    stringsAsFactors = FALSE,
    colClasses = "character",
    na.strings = character(),
    fileEncoding = "UTF-8-BOM"
  )
  if (anyDuplicated(names(data))) fail("input contains duplicate column names")
  required <- c("id", "group", schema$variable)
  missing <- setdiff(required, names(data))
  if (length(missing)) fail(paste0("input is missing schema field(s): ", paste(missing, collapse = ", ")))
  if (!nrow(data)) fail("input contains no subjects")
  data$id <- trimws(data$id)
  data$group <- trimws(data$group)
  if (any(!nzchar(data$id)) || any(!nzchar(data$group))) fail("id and group must be non-empty")
  if (anyDuplicated(data$id)) fail("id must be unique: one input row per subject")
  for (index in seq_len(nrow(schema))) {
    variable <- schema$variable[[index]]
    text <- trimws(data[[variable]])
    missing_value <- !nzchar(text)
    if (schema$type[[index]] == "continuous") {
      value <- suppressWarnings(as.numeric(text))
      bad <- !missing_value & (is.na(value) | !is.finite(value))
      if (any(bad)) fail(paste0(variable, " must contain finite numeric values or blanks"))
      value[missing_value] <- NA_real_
      data[[variable]] <- value
    } else {
      bad <- !missing_value & !text %in% schema$level_list[[index]]
      if (any(bad)) fail(paste0(variable, " contains a category not declared in schema levels"))
      text[missing_value] <- NA_character_
      data[[variable]] <- text
    }
  }
  data
}

make_columns <- function(data, include_overall) {
  group_order <- ordered_unique(data$group)
  columns <- lapply(group_order, function(group_name) data[data$group == group_name, , drop = FALSE])
  names(columns) <- group_order
  if (include_overall) {
    if ("Overall" %in% group_order) fail("group label 'Overall' conflicts with --include-overall true")
    columns[["Overall"]] <- data
  }
  if (length(columns) > MAX_GROUP_COLUMNS) {
    fail(paste0(
      "table would have ", length(columns), " group columns; maximum is ", MAX_GROUP_COLUMNS,
      ". Split the table using a scientifically defined grouping."
    ))
  }
  columns
}

format_number <- function(value, decimals) sprintf(paste0("%.", decimals, "f"), value)

build_rows <- function(schema, columns) {
  rows <- list()
  for (index in seq_len(nrow(schema))) {
    variable <- schema$variable[[index]]
    kind <- schema$type[[index]]
    availability <- vapply(columns, function(frame) {
      available <- sum(!is.na(frame[[variable]]))
      sprintf("available %d/%d; missing %d", available, nrow(frame), nrow(frame) - available)
    }, character(1L))
    rows[[length(rows) + 1L]] <- list(kind = "section", label = schema$label[[index]], cells = availability)
    if (kind == "continuous") {
      label <- if (schema$summary[[index]] == "mean_sd") "Mean (SD)" else "Median [Q1, Q3]"
      cells <- vapply(columns, function(frame) {
        values <- frame[[variable]][!is.na(frame[[variable]])]
        if (!length(values)) return("NA")
        if (schema$summary[[index]] == "mean_sd") {
          sd_text <- if (length(values) > 1L) format_number(sd(values), schema$decimals[[index]]) else "NA"
          paste0(format_number(mean(values), schema$decimals[[index]]), " (", sd_text, ")")
        } else {
          quartiles <- quantile(values, c(0.25, 0.5, 0.75), type = 7, names = FALSE)
          paste0(
            format_number(quartiles[[2L]], schema$decimals[[index]]), " [",
            format_number(quartiles[[1L]], schema$decimals[[index]]), ", ",
            format_number(quartiles[[3L]], schema$decimals[[index]]), "]"
          )
        }
      }, character(1L))
      rows[[length(rows) + 1L]] <- list(kind = "data", label = label, cells = cells)
    } else {
      for (level in schema$level_list[[index]]) {
        cells <- vapply(columns, function(frame) {
          values <- frame[[variable]][!is.na(frame[[variable]])]
          count <- sum(values == level)
          if (length(values)) sprintf("%d/%d (%.1f%%)", count, length(values), 100 * count / length(values)) else "0/0 (NA)"
        }, character(1L))
        rows[[length(rows) + 1L]] <- list(kind = "data", label = level, cells = cells)
      }
    }
  }
  if (length(rows) > MAX_DISPLAY_ROWS) {
    fail(paste0(
      "schema expands to ", length(rows), " display rows; maximum is ", MAX_DISPLAY_ROWS,
      ". Split variables into themed tables rather than shrinking text."
    ))
  }
  rows
}

geometry <- function(rows, columns) {
  maximum_label <- max(nchar(vapply(rows, function(row) row$label, character(1L)), type = "width"))
  list(
    width = min(20, max(8.5, 4.6 + 2.1 * length(columns) + 0.045 * maximum_label)),
    height = max(5, 2.4 + 0.42 * (length(rows) + 1L)),
    maximum_label = maximum_label
  )
}

draw_table <- function(data, title_text, columns, rows, size_info) {
  par(mar = c(0, 0, 0, 0), family = "sans", bg = "white")
  plot.new()
  plot.window(xlim = c(0, 1), ylim = c(0, 1), xaxs = "i", yaxs = "i")
  colors <- c(ink = "#17212B", muted = "#617080", header = "#17354D", section = "#EAF3F2", stripe = "#F7F9FA", line = "#D8E0E5", white = "#FFFFFF")
  left <- 0.035
  right <- 0.975
  top <- 0.865
  bottom <- 0.105
  label_fraction <- min(0.44, max(0.30, 0.28 + 0.004 * size_info$maximum_label))
  label_right <- left + (right - left) * label_fraction
  data_width <- (right - label_right) / length(columns)
  edges <- c(left, label_right, label_right + data_width * seq_len(length(columns)))
  weights <- c(1.28, vapply(rows, function(row) if (row$kind == "section") 1.05 else 1, numeric(1L)))
  unit_height <- (top - bottom) / sum(weights)

  text(left, 0.962, title_text, adj = c(0, 1), cex = 1.55, font = 2, col = colors["ink"])
  text(
    left, 0.920,
    sprintf("%d subject rows · %d displayed group columns · descriptive summaries only", nrow(data), length(columns)),
    adj = c(0, 1), cex = 0.82, col = colors["muted"]
  )
  y <- top
  header_height <- weights[[1L]] * unit_height
  rect(left, y - header_height, right, y, col = colors["header"], border = NA)
  text(left + 0.012, y - header_height / 2, "Variable / summary", adj = c(0, 0.5), cex = 0.86, font = 2, col = "white")
  for (index in seq_along(columns)) {
    center <- mean(edges[c(index + 1L, index + 2L)])
    text(center, y - header_height / 2, paste0(names(columns)[[index]], "\nN=", nrow(columns[[index]])), cex = 0.80, font = 2, col = "white")
  }
  y <- y - header_height
  data_index <- 0L
  for (index in seq_along(rows)) {
    row <- rows[[index]]
    row_height <- weights[[index + 1L]] * unit_height
    fill <- if (row$kind == "section") colors["section"] else if (data_index %% 2L) colors["stripe"] else colors["white"]
    rect(left, y - row_height, right, y, col = fill, border = NA)
    segments(left, y - row_height, right, y - row_height, col = colors["line"], lwd = 0.65)
    text(
      left + if (row$kind == "section") 0.012 else 0.022,
      y - row_height / 2,
      row$label,
      adj = c(0, 0.5),
      cex = 0.79,
      font = if (row$kind == "section") 2 else 1,
      col = colors["ink"]
    )
    for (column_index in seq_along(row$cells)) {
      center <- mean(edges[c(column_index + 1L, column_index + 2L)])
      text(
        center, y - row_height / 2, row$cells[[column_index]],
        cex = if (row$kind == "section") 0.64 else 0.75,
        col = if (row$kind == "section") colors["muted"] else colors["ink"]
      )
    }
    if (row$kind == "data") data_index <- data_index + 1L
    y <- y - row_height
  }
  rect(left, bottom, right, top, border = colors["header"], lwd = 1)
  text(
    left, 0.060,
    "Categorical percentages use each variable's non-missing denominator. Continuous summaries exclude missing values. SD is sample SD; quartiles use type 7. No P values or SMD are computed.",
    adj = c(0, 1), cex = 0.67, col = colors["muted"]
  )
}

render_output <- function(kind, path, data, title_text, columns, rows, size_info) {
  if (kind == "png") {
    png(path, width = size_info$width, height = size_info$height, units = "in", res = 320, type = if (capabilities("cairo")) "cairo" else getOption("bitmapType"), bg = "white")
  } else {
    svg(path, width = size_info$width, height = size_info$height, onefile = TRUE, bg = "white", family = "sans")
  }
  device_id <- dev.cur()
  on.exit(if (dev.cur() == device_id) dev.off(), add = FALSE)
  draw_table(data, title_text, columns, rows, size_info)
  dev.off()
  on.exit(NULL, add = FALSE)
}

report_data <- function(data, schema, columns) {
  missing_cells <- sum(vapply(schema$variable, function(variable) sum(is.na(data[[variable]])), integer(1L)))
  cat(sprintf(
    "Loaded %d unique subject rows: %d displayed group column(s), %d typed variable(s), %d missing value cell(s); 0 subject rows excluded.\n",
    nrow(data), length(columns), nrow(schema), missing_cells
  ))
  for (index in seq_len(nrow(schema))) {
    variable <- schema$variable[[index]]
    cat(sprintf("variable=%s | type=%s | summary=%s\n", variable, schema$type[[index]], schema$summary[[index]]))
    for (group_name in names(columns)) {
      frame <- columns[[group_name]]
      available <- sum(!is.na(frame[[variable]]))
      cat(sprintf("  group=%s | N=%d | available=%d | missing=%d\n", group_name, nrow(frame), available, nrow(frame) - available))
    }
  }
}

main <- function() {
  config <- parse_args(commandArgs(trailingOnly = TRUE))
  schema <- read_schema(config$schema)
  data <- read_data(config$input, schema)
  columns <- make_columns(data, config$include_overall)
  rows <- build_rows(schema, columns)
  size_info <- geometry(rows, columns)
  output_directory <- dirname(config$output_prefix)
  if (!dir.exists(output_directory)) dir.create(output_directory, recursive = TRUE, showWarnings = FALSE)
  png_path <- paste0(config$output_prefix, ".png")
  svg_path <- paste0(config$output_prefix, ".svg")
  render_output("png", png_path, data, config$title, columns, rows, size_info)
  render_output("svg", svg_path, data, config$title, columns, rows, size_info)
  report_data(data, schema, columns)
  cat("Wrote ", png_path, "\n", sep = "")
  cat("Wrote ", svg_path, "\n", sep = "")
}

status <- tryCatch({main(); 0L}, error = function(error) {message("ERROR: ", conditionMessage(error)); 2L})
quit(save = "no", status = status)
